### PR TITLE
Add `override` keyword to compile with `-Wsuggest-override`

### PR DESCRIPTION
**Pull request by alexreinking**
The `override` keyword is a C++11 feature that helps document which functions are actually overriding base class virtual functions. By requiring this keyword with `-Wsuggest-override` (on g++>=5.1 only) we can avoid accidentally overriding a base class method and can quickly tell when methods are overriding others.

### CODE REVIEW

**Comment by steven-johnson** [comment]
IIRC @abadams has opinions on this, so opening to him for comments :-)

File: CMakeLists.txt [hunk 1]
```diff
@@ -206,11 +206,14 @@ else()
...
-                        -Wcast-qual
-                        -Woverloaded-virtual
-                        -Wignored-qualifiers)
+  add_compile_options(-Wall
+                      -Wno-unused-function
+                      -Wcast-qual
+                      -Woverloaded-virtual
+                      -Wignored-qualifiers)
+  if (CMAKE_COMPILER_IS_GNUCC AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER 5.1)
+    add_compile_options(-Wsuggest-override)
```
Comment (steven-johnson) [inline CMakeLists.txt]: Does this handle Clang as well?

**Comment by abadams** [comment]
Yes, this sort of thing could break the build somewhere.

File: CMakeLists.txt (hunk 1 above)
Comment (alexreinking) [reply CMakeLists.txt]: @steven-johnson Clang has [`-Winconsistent-missing-override`](https://clang.llvm.org/docs/DiagnosticsReference.html#winconsistent-missing-override) enabled by default, and `-Wsuggest-override` is G++-specific.

**Comment by alexreinking** [comment]
Fixing version check in Makefile to support g++ 5.1 - 6.0

**Comment by steven-johnson** [comment]
Cloned to https://github.com/halide/Halide/pull/3343 for testing

**Comment by steven-johnson** [comment]
Cloned PR passed, landing

**Pull request merged**